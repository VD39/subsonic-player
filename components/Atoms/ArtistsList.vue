<script setup lang="ts">
defineProps<{
  artists: BaseArtist[];
}>();
</script>

<template>
  <ul class="commaList">
    <li
      v-for="artist in artists"
      :key="artist.id"
      data-test-id="artist-list-item"
    >
      <NuxtLink :to="`/artist/${artist.id}`" class="link">
        {{ artist.name }}
      </NuxtLink>
    </li>
  </ul>
</template>
