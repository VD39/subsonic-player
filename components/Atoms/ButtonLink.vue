<script setup lang="ts">
const props = withDefaults(
  defineProps<{
    disabled?: boolean;
    fullWidth?: boolean;
    icon?: Component | Icon;
    iconColor?: string;
    iconPosition?: IconPosition;
    iconSize?: IconSize;
    iconWeight?: IconWeight;
    is?: ButtonLink;
    showText?: boolean;
    type?: string;
  }>(),
  {
    disabled: undefined,
    icon: undefined,
    iconColor: undefined,
    iconPosition: 'left',
    iconSize: 'medium',
    iconWeight: 'regular',
    is: 'button',
    type: 'button',
  },
);

const type = props.is === 'button' ? props.type : undefined;
const isComponent =
  props.is === 'nuxt-link' ? resolveComponent('NuxtLink') : props.is;
</script>

<template>
  <component
    :is="isComponent"
    :class="[
      'centerItems',
      $style.buttonLink,
      {
        [$style.alignRight]: icon && iconPosition === 'right',
        [$style.disabled]: disabled,
        'centerAll fullWidth': fullWidth,
      },
    ]"
    :disabled
    draggable="false"
    :type
    v-bind="$attrs"
  >
    <component
      :is="icon"
      v-if="icon"
      ref="iconComponent"
      aria-hidden="true"
      :class="$style.icon"
      :color="iconColor"
      :size="ICON_SIZE[iconSize]"
      :weight="iconWeight"
    />

    <span
      ref="text"
      :class="[
        'clamp',
        $style.text,
        {
          visuallyHidden: !showText,
        },
      ]"
    >
      <slot />
    </span>
  </component>
</template>

<style module>
.buttonLink {
  position: relative;
  gap: var(--default-space);
  padding: var(--default-space);
  white-space: nowrap;
}

.alignRight {
  flex-direction: row-reverse;
}

.disabled {
  pointer-events: none;
  opacity: 0.25;
}

.icon {
  flex-shrink: 0;

  [opacity='0.2'] {
    opacity: 0.4;
    fill: var(--secondary-font-color);
  }
}

.text {
  display: block;
  margin-bottom: 1px;
}
</style>
