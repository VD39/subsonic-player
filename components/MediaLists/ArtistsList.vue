<script setup lang="ts">
import NoMediaMessage from '@/components/NoMediaMessage/NoMediaMessage.vue';
import MediaListWrapper from './MediaListWrapper.vue';
import ArtistItem from './ArtistItem.vue';

defineProps<{
  artists: Artist[];
  rows?: string;
}>();
</script>

<template>
  <MediaListWrapper v-if="artists.length" :row="rows">
    <ArtistItem v-for="artist in artists" :key="artist.id" :artist="artist" />
  </MediaListWrapper>
  <NoMediaMessage v-else icon="PhUsersThree" message="No artists found." />
</template>
