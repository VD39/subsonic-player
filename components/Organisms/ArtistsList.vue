<script setup lang="ts">
import MediaListWrapper from '@/components/Atoms/MediaListWrapper.vue';
import NoMediaMessage from '@/components/Atoms/NoMediaMessage.vue';
import ArtistItem from '@/components/Organisms/ArtistItem.vue';

defineProps<{
  artists: Artist[] | SimilarArtist[];
}>();
</script>

<template>
  <MediaListWrapper v-if="artists.length">
    <ArtistItem v-for="artist in artists" :key="artist.id" :artist />
  </MediaListWrapper>

  <NoMediaMessage
    v-else
    :icon="IMAGE_DEFAULT_BY_TYPE.artist"
    message="No artists found."
  />
</template>
