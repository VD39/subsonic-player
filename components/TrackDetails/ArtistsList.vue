<script setup lang="ts">
defineProps<{
  artists: BaseArtist[];
}>();
</script>

<template>
  <ul :class="[$style.meta, $style.list]">
    <li
      v-for="artist in artists"
      :key="artist.id"
      :class="$style.listItem"
      data-test-id="artist-list-item"
    >
      <NuxtLink :to="`/artist/${artist.id}`">
        {{ artist.name }}
      </NuxtLink>
    </li>
  </ul>
</template>

<style module>
.list {
  @mixin align-center;
}

.listItem {
  &:not(:last-child)::after {
    content: ',\00a0';
  }

  &:nth-last-child(2)::after {
    content: '\00a0&\00a0';
  }
}
</style>
