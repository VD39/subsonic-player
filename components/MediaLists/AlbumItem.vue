<script setup lang="ts">
import ArtistsList from '@/components/TrackDetails/ArtistsList.vue';
import ImageLink from '@/components/Image/ImageLink.vue';

const props = defineProps<{
  album: Album;
}>();

const toLink = `/album/${props.album.id}`;
</script>

<template>
  <article>
    <ImageLink :to="toLink" :title="album.name" :image="album.image" />

    <p class="clamp clamp2">
      <NuxtLink :to="toLink" :title="`Go to album ${album.name}`">
        {{ album.name }}
      </NuxtLink>
    </p>

    <ArtistsList
      v-if="album.artists.length"
      :artists="album.artists"
      class="clamp clamp2"
    />
  </article>
</template>
