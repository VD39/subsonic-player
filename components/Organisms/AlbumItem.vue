<script setup lang="ts">
import ArtistsList from '@/components/Atoms/ArtistsList.vue';
import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import SpinningLoader from '@/components/Atoms/SpinningLoader.vue';
import ImageLink from '@/components/Organisms/ImageLink.vue';

const props = defineProps<{
  album: Album;
  hideArtist?: boolean;
}>();

const { playTracks } = useAudioPlayer();
const { getAlbum } = useAlbum();

const loading = ref(false);

async function playAlbumTracks(albumId: string) {
  loading.value = true;

  const selectedAlbum = await getAlbum(albumId);

  if (selectedAlbum) {
    await playTracks(selectedAlbum.tracks);
  }

  loading.value = false;
}

const buttonProps = computed<ButtonProps>(() => ({
  icon: loading.value ? SpinningLoader : ICONS.play,
  text: `Play album ${props.album.name}`,
}));

const toLink = `/album/${props.album.id}`;
</script>

<template>
  <article>
    <div :class="$style.albumImageWrapper">
      <ImageLink
        :to="toLink"
        :title="`Go to album ${album.name}`"
        :image="album.image"
        class="mBS"
      />

      <ButtonLink
        :icon="buttonProps.icon"
        :class="['themeHoverButton', $style.buttonLink]"
        :title="buttonProps.text"
        @click="playAlbumTracks(album.id)"
      >
        {{ buttonProps.text }}
      </ButtonLink>
    </div>

    <p class="mBXS strong smallFont clamp2">
      <NuxtLink
        :to="toLink"
        class="link"
        :aria-label="`Go to album ${album.name}`"
      >
        {{ album.name }}
      </NuxtLink>
    </p>

    <ArtistsList
      v-if="!hideArtist && album.artists.length"
      :artists="album.artists"
      class="smallFont clamp2"
    />
  </article>
</template>

<style module>
.albumImageWrapper {
  position: relative;

  @media (hover: hover) {
    &:hover,
    &:focus {
      .buttonLink {
        --album-play-button-opacity: 1;
      }
    }
  }
}

.buttonLink {
  --album-play-button-opacity: 0;

  position: absolute;
  inset: auto var(--default-space) var(--default-space) auto;
  opacity: var(--album-play-button-opacity);
  transition: opacity var(--transition);
}
</style>
