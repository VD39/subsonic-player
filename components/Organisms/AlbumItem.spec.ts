import type { VueWrapper } from '@vue/test-utils';

import ArtistsList from '@/components/Atoms/ArtistsList.vue';
import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import { formattedAlbumMock } from '@/test/fixtures';
import { useAudioPlayerMock } from '@/test/useAudioPlayerMock';
import { mockNuxtImport } from '@nuxt/test-utils/runtime';
import { mount } from '@vue/test-utils';

import AlbumItem from './AlbumItem.vue';

const albumMock = ref<Album | null>(null);
const getAlbumMock = vi.fn();

mockNuxtImport('useAlbum', () => () => ({
  album: albumMock,
  getAlbum: getAlbumMock,
}));

const { playTracksMock } = useAudioPlayerMock();

function factory(props = {}) {
  return mount(AlbumItem, {
    props: {
      album: formattedAlbumMock,
      hideArtist: false,
      ...props,
    },
  });
}

describe('AlbumItem', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('when album artist is not an empty array', () => {
    describe('when hideArtist prop is false', () => {
      it('shows the ArtistsList component', () => {
        expect(wrapper.findComponent(ArtistsList).exists()).toBe(true);
      });
    });

    describe('when hideArtist prop is true', () => {
      beforeEach(() => {
        wrapper = factory({
          hideArtist: true,
        });
      });

      it('does not show the ArtistsList component', () => {
        expect(wrapper.findComponent(ArtistsList).exists()).toBe(false);
      });
    });
  });

  describe('when album artist is an empty array', () => {
    beforeEach(() => {
      wrapper = factory({
        album: {
          ...formattedAlbumMock,
          artists: [],
        },
      });
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('does not show the ArtistsList component', () => {
      expect(wrapper.findComponent(ArtistsList).exists()).toBe(false);
    });
  });

  describe('when ButtonLink component is clicked', () => {
    beforeEach(async () => {
      await wrapper.findComponent(ButtonLink).trigger('click');
    });

    it('calls the getAlbum function', () => {
      expect(getAlbumMock).toHaveBeenCalledWith(formattedAlbumMock.id, true);
    });

    describe('when useAlbum album returns null', () => {
      it('does not call the playTracks function', () => {
        expect(playTracksMock).not.toHaveBeenCalled();
      });
    });

    describe('when useAlbum album returns track data', () => {
      beforeEach(async () => {
        albumMock.value = formattedAlbumMock;
        await wrapper.findComponent(ButtonLink).trigger('click');
      });

      it('calls the playTracks function', () => {
        expect(playTracksMock).toHaveBeenCalledWith(formattedAlbumMock.tracks);
      });
    });
  });
});
