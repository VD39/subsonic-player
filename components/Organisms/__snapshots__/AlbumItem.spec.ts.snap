// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AlbumItem > matches the snapshot 1`] = `
"<article>
  <div class="albumImageWrapper">
    <figure class="mBS"><a draggable="false" title="Go to album album-0">
        <div class="centerAll preloadImage"><span class="skeletonLoader"><span class="visuallyHidden">Loading image</span></span><img alt="Go to album album-0" class="image" draggable="false" loading="lazy" src="image" style="display: none;"></div>
      </a></figure>
    <div class="actions"><button class="centerItems buttonLink" draggable="false" type="button" title="Unlike album">
        <phheart aria-hidden="true" class="icon" color="var(--error-color)" size="24" weight="fill"></phheart><span class="clamp text visuallyHidden">Unlike album</span>
      </button><button class="centerItems buttonLink themeHoverButton themeHoverButton" draggable="false" type="button" title="Play album album-0">
        <phplay aria-hidden="true" class="icon" size="24" weight="regular"></phplay><span class="clamp text visuallyHidden">Play album album-0</span>
      </button></div>
  </div>
  <p class="mBXS strong smallFont clamp2"><a aria-label="Go to album album-0" class="link" draggable="false">album-0</a></p>
  <ul class="commaList smallFont clamp2">
    <li data-test-id="artist-list-item"><a aria-label="Go to artist name" class="link" draggable="false">name</a></li>
    <li data-test-id="artist-list-item"><a aria-label="Go to artist name1" class="link" draggable="false">name1</a></li>
  </ul>
</article>"
`;

exports[`AlbumItem > when album artist is an empty array > matches the snapshot 1`] = `
"<article>
  <div class="albumImageWrapper">
    <figure class="mBS"><a draggable="false" title="Go to album album-0">
        <div class="centerAll preloadImage"><span class="skeletonLoader"><span class="visuallyHidden">Loading image</span></span><img alt="Go to album album-0" class="image" draggable="false" loading="lazy" src="image" style="display: none;"></div>
      </a></figure>
    <div class="actions"><button class="centerItems buttonLink" draggable="false" type="button" title="Unlike album">
        <phheart aria-hidden="true" class="icon" color="var(--error-color)" size="24" weight="fill"></phheart><span class="clamp text visuallyHidden">Unlike album</span>
      </button><button class="centerItems buttonLink themeHoverButton themeHoverButton" draggable="false" type="button" title="Play album album-0">
        <phplay aria-hidden="true" class="icon" size="24" weight="regular"></phplay><span class="clamp text visuallyHidden">Play album album-0</span>
      </button></div>
  </div>
  <p class="mBXS strong smallFont clamp2"><a aria-label="Go to album album-0" class="link" draggable="false">album-0</a></p>
  <!--v-if-->
</article>"
`;
